### PR TITLE
[Feature] Support Generated Column for Olap Table

### LINKED ISSUE #20305: [Feature] Support Materialized Column
Materialized Column is a kind of column that can evaluated through other columns in the table. The expensive computation can be done when loading the data and speed up the query using this technique.

usage:
1. using CREATE TABLE:

CREATE TABLE t ( id BIGINT NOT NULL,  array_int ARRAY<int> NOT NULL, array_string ARRAY<string> NOT NULL, json_data JSON NOT NULL,
mc_1 DOUBLE MATERIALIZED_COLUMN (array_avg(array_int)),
mc_2 ARRAY<int> MATERIALIZED_COLUMN (array_sort(array_int)),
mc_3 ARRAY<string> MATERIALIZED_COLUMN (array_sort(array_string)),
mc_4 String MATERIALIZED_COLUMN (get_json_string(json_string(json_data), "$.a")) )
Primary KEY (id) DISTRIBUTED BY HASH(id) BUCKETS 48 PROPERTIES("replication_num" = "1", "replicated_storage"="true");

2. using ALTER TABLE:
ALTER TABLE t ADD MATERIALIZED COLUMN mc_5 BIGINT (id);

### PR DESCRIPTION
**Pull request by srlch**
## What type of PR is this：
- [ ] BugFix
- [x] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

## Which issues of this PR fixes ：

Fixes #20305

## Problem Summary(Required) ：
Materialized Column is a kind of column that can evaluated through other columns in the table.
The expensive computation can be done when loading the data and speed up the query using this
technique.

usage:
1. using CREATE TABLE:

```
CREATE TABLE t (
  id BIGINT NOT NULL,  
  array_int ARRAY<int> NOT NULL,
  array_string ARRAY<string> NOT NULL,
  json_data JSON NOT NULL,
  mc_1 DOUBLE AS array_avg(array_int),
  mc_2 ARRAY<int> AS array_sort(array_int),
  mc_3 ARRAY<string> AS array_sort(array_string),
  mc_4 String AS get_json_string(json_string(json_data), $.a)
)
Primary KEY (id)
DISTRIBUTED BY HASH(id)
BUCKETS 48;
```

2. using ALTER TABLE: ALTER TABLE t ADD COLUMN mc_5 BIGINT AS (id);

## Checklist:

- [x] I have added test cases for my bug fix or my new feature
- [ ] This pr will affect users' behaviors
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto backported to target branch
  - [ ] 3.0
  - [ ] 2.5
  - [ ] 2.4
  - [ ] 2.3

### CODE REVIEW

File: be/src/storage/schema_change.cpp [hunk 1]
```diff
@@ -40,6 +40,8 @@
 #include <vector>
 
 #include "exec/sorting/sorting.h"
+#include "exprs/expr.h"
+ #include "exprs/expr_context.h"
```
Comment (chaoyli) [inline be/src/storage/schema_change.cpp:44]: The two lines is not aligned.

File: fe/fe-core/src/main/java/com/starrocks/sql/ast/AddMaterializedColumnClause.java [hunk 2]
```diff
@@ -0,0 +1,45 @@
+// Copyright 2021-present StarRocks, Inc. All rights reserved.
+//
```
Comment (chaoyli) [inline fe/fe-core/src/main/java/com/starrocks/sql/ast/AddMaterializedColumnClause.java:2]: What expressions do you provide for the materialized column? Does the SchemaChangeJobV2 in BE support the get_json_string?

File: fe/fe-core/src/main/java/com/starrocks/sql/ast/DropColumnClause.java [hunk 3]
```diff
@@ -23,18 +23,25 @@
 // Drop one column
 public class DropColumnClause extends AlterTableColumnClause {
     private final String colName;
+    private final boolean materializedColumn;
```
Comment (chaoyli) [inline fe/fe-core/src/main/java/com/starrocks/sql/ast/DropColumnClause.java:26]:
```suggestion
    private final boolean isMaterializedColumn;
```

File: fe/fe-core/src/main/java/com/starrocks/sql/ast/ModifyMaterializedColumnClause.java [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+
+
+package com.starrocks.sql.ast;
+
+import com.starrocks.analysis.ColumnDef;
+import com.starrocks.analysis.Expr;
+import com.starrocks.sql.parser.NodePosition;
+
+// modify one materialized column
+public class ModifyMaterializedColumnClause extends ModifyColumnClause {
```
Comment (chaoyli) [inline fe/fe-core/src/main/java/com/starrocks/sql/ast/ModifyMaterializedColumnClause.java:23]: `alter table modify materialized column`? Under which case, this is necessary?

File: fe/fe-core/src/main/java/com/starrocks/sql/parser/StarRocks.g4 [hunk 5]
```diff
@@ -780,8 +786,17 @@ addColumnsClause
...
 
+addMaterializedColumnClause
+    : ADD MATERIALIZED COLUMN columnDesc '(' expression ')'
+    ;
+
+modifyMaterializedColumnClause
+    : MODIFY MATERIALIZED COLUMN columnDesc ('(' expression ')')?
+    ;
+
 dropColumnClause
```
Comment (chaoyli) [inline fe/fe-core/src/main/java/com/starrocks/sql/parser/StarRocks.g4:797]: If we drop the base column, what behavior will happen?

File: be/src/storage/schema_change_utils.cpp [hunk 6]
```diff
@@ -438,6 +450,34 @@ bool ChunkChanger::change_chunk_v2(ChunkPtr& base_chunk, ChunkPtr& new_chunk, co
     return true;
 }
 
+Status ChunkChanger::fill_materialized_columns(ChunkPtr& new_chunk) {
```
Comment (chaoyli) [inline be/src/storage/schema_change_utils.cpp:453]: `add column` will enter the linked schema change, it does not change the historical data.
I think it is not suitable for `add materialized column`, it should change the historical data.

File: fe/fe-core/src/main/java/com/starrocks/alter/MaterializedViewHandler.java [hunk 7]
```diff
@@ -351,6 +351,13 @@ private RollupJobV2 createMaterializedViewJob(String mvName, String baseIndexNam
         // get timeout
         long timeoutMs = PropertyAnalyzer.analyzeTimeout(properties, Config.alter_table_timeout_second) * 1000;
 
+        for (Column col : mvColumns) {
+            Column originColumn = olapTable.getColumn(col.getName());
+            if (originColumn.isMaterializedColumn()) {
+                throw new DdlException("Do not support create materialized view with Materialized Column");
```
Comment (HangyuanLiu) [inline fe/fe-core/src/main/java/com/starrocks/alter/MaterializedViewHandler.java:357]: Semantic Check move to Analyzer

File: fe/fe-core/src/main/java/com/starrocks/sql/parser/StarRocks.g4 [hunk 8]
```diff
@@ -687,6 +691,8 @@ alterClause
     | columnRenameClause
     | reorderColumnsClause
     | rollupRenameClause
+    | addMaterializedColumnClause
```
Comment (HangyuanLiu) [inline fe/fe-core/src/main/java/com/starrocks/sql/parser/StarRocks.g4:694]: Merge into existing grammars, no need to add grammars separately. Same as when creating a table, modify columnDesc uniformly

File: fe/fe-core/src/main/java/com/starrocks/alter/Alter.java [hunk 9]
```diff
@@ -138,6 +138,10 @@ public void start() {
         clusterHandler.start();
     }
 
+    public AlterHandler schemaChangeHandler() {
```
Comment (HangyuanLiu) [inline fe/fe-core/src/main/java/com/starrocks/alter/Alter.java:141]: getSchemaChangeHandler has exists

File: fe/fe-core/src/main/java/com/starrocks/alter/SchemaChangeJobV2.java [hunk 10]
```diff
@@ -489,6 +508,79 @@ protected void runWaitingTxnJob() throws AlterCancelException {
...
+                                                                "the ref column");
+                                }
+
+                                smap.getLhs().add(slot);
+                                SlotRef slotRef = new SlotRef(slotDesc);
+                                slotRef.setColumnName(slot.getColumnName());
+                                smap.getRhs().add(slotRef);
+                            }
+
+                            Expr materializedColumnExpr = expr.clone(smap);
```
Comment (HangyuanLiu) [inline fe/fe-core/src/main/java/com/starrocks/alter/SchemaChangeJobV2.java:555]: What is the purpose of this logic?

File: fe/fe-core/src/main/java/com/starrocks/server/OlapTableFactory.java [hunk 11]
```diff
@@ -542,6 +556,75 @@ public Table createTable(LocalMetastore metastore, Database db, CreateTableStmt
...
+                    }
+
+                    if (column.isMaterializedColumn()) {
+                        Expr expr = column.materializedColumnExpr();
+                        TableName tblName = new TableName(db.getFullName(), table.getName());
+
+                        ConnectContext connectContext = new ConnectContext();
+                        connectContext.setDatabase(db.getFullName());
+                        // set privilege
+                        connectContext.setQualifiedUser(AuthenticationManager.ROOT_USER);
```
Comment (HangyuanLiu) [inline fe/fe-core/src/main/java/com/starrocks/server/OlapTableFactory.java:585]: Why need set privilege?

File: be/src/storage/schema_change_utils.cpp (hunk 6 above)
Comment (chaoyli) [reply be/src/storage/schema_change_utils.cpp:453]: 2. `mc_1 DOUBLE MATERIALIZED_COLUMN (array_avg(array_int))` this syntax seems a little strange.
3. What functions are supported in MATERIALIZED_COLUMN?

File: fe/fe-core/src/main/java/com/starrocks/alter/SchemaChangeHandler.java [hunk 12]
```diff
@@ -154,6 +155,18 @@ private void processAddColumn(AddColumnClause alterClause, OlapTable olapTable,
         Set<String> newColNameSet = Sets.newHashSet(column.getName());
         addColumnInternal(olapTable, column, columnPos, targetIndexId, baseIndexId,
                 indexSchemaMap, newColNameSet);
+
+        List<Column> alterMaterializedColumns = Lists.newArrayList();
+        if (alterClause.materializedColumnExpr() != null) {
```
Comment (HangyuanLiu) [inline fe/fe-core/src/main/java/com/starrocks/alter/SchemaChangeHandler.java:160]: Use the materialize column information in Column?

File: fe/fe-core/src/main/java/com/starrocks/analysis/KeysDesc.java [hunk 13]
```diff
@@ -80,6 +80,10 @@ public int keysColumnSize() {
         return keysColumnNames.size();
     }
 
+    public List<String> getKeysColumnNames() {
```
Comment (HangyuanLiu) [inline fe/fe-core/src/main/java/com/starrocks/analysis/KeysDesc.java:83]: use containsCol

File: be/src/storage/schema_change.cpp [hunk 14]
```diff
@@ -308,6 +310,11 @@ Status SchemaChangeDirectly::process(TabletReader* reader, RowsetWriter* new_row
             return Status::InternalError(err_msg);
         }
 
+        if (auto st = _chunk_changer->fill_materialized_columns(new_chunk); !st.ok()) {
+            LOG(WARNING) << "fill materialized columns failed: " << st;
```
Comment (fzhedu) [inline be/src/storage/schema_change.cpp:314]: `        LOG(WARNING) << "fill materialized columns failed: " << st.get_error_msg();
`

File: be/src/storage/schema_change.cpp [hunk 15]
```diff
@@ -330,6 +337,10 @@ Status SchemaChangeDirectly::process(TabletReader* reader, RowsetWriter* new_row
             LOG(WARNING) << err_msg;
             return Status::InternalError(err_msg);
         }
+        if (auto st = _chunk_changer->fill_materialized_columns(new_chunk); !st.ok()) {
+            LOG(WARNING) << "fill materialized columns failed: " << st;
```
Comment (fzhedu) [inline be/src/storage/schema_change.cpp:341]: ditto

**Review by chaoyli — DISMISSED** [review]

File: be/src/storage/schema_change_utils.cpp [hunk 16]
```diff
@@ -438,6 +450,40 @@ bool ChunkChanger::change_chunk_v2(ChunkPtr& base_chunk, ChunkPtr& new_chunk, co
...
+            // Only null column maybe lost type info, we append null
+            // for the chunk instead of swapping the tmp column.
+            std::dynamic_pointer_cast<NullableColumn>(new_chunk->get_column_by_index(it.first))->reset_column();
+            std::dynamic_pointer_cast<NullableColumn>(new_chunk->get_column_by_index(it.first))
+                    ->append_nulls(new_chunk->num_rows());
+        } else if (tmp->is_nullable()) {
+            new_chunk->get_column_by_index(it.first).swap(tmp);
+        } else {
+            // materialized column must be a nullable column. If tmp is not nullable column,
+            // new_chunk can not swap it directly
```
Comment (fzhedu) [inline be/src/storage/schema_change_utils.cpp:475]: judge this anywhere?

Comment (fzhedu) [reply be/src/storage/schema_change_utils.cpp:475]: otherwise, the following cast may cause null.

File: be/src/storage/schema_change_utils.h [hunk 17]
```diff
@@ -42,12 +43,22 @@ class ChunkChanger {
 
     const std::vector<ColumnId>& get_selected_column_indexes() const { return _selected_column_indexes; }
 
+    ObjectPool* get_object_pool() { return &_obj_pool; }
+
```
Comment (fzhedu) [inline be/src/storage/schema_change_utils.h:47]: add const after function?

File: be/src/storage/schema_change_utils.h [hunk 18]
```diff
@@ -42,12 +43,22 @@ class ChunkChanger {
...
+    RuntimeState* get_runtime_state() { return _state; }
```
Comment (fzhedu) [inline be/src/storage/schema_change_utils.h:48]: diitto

File: fe/fe-core/src/main/java/com/starrocks/alter/SchemaChangeJobV2.java [hunk 19]
```diff
@@ -491,6 +516,96 @@ protected void runWaitingTxnJob() throws AlterCancelException {
...
+                                    column.isNameWithPrefix(SchemaChangeHandler.SHADOW_NAME_PRFIX_V1)) {
+                                String originName = Column.removeNamePrefix(column.getName());
+                                columnIndex = tbl.getFullSchema().indexOf(tbl.getColumn(originName));
+                            } else {
+                                columnIndex = tbl.getFullSchema().indexOf(column);
+                            }
+
+                            mcExprs.put(columnIndex, materializedColumnExpr.treeToThrift());
+
+                            // we need this thing, otherwise, otherwise some expr evalution will fail in BE
```
Comment (fzhedu) [inline fe/fe-core/src/main/java/com/starrocks/alter/SchemaChangeJobV2.java:595]: typo

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 183 / 477 (38.36%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/ast/AddColumnClause.java|0|2|00.00%|[35, 36]|
|:large_blue_circle:|com/starrocks/alter/SchemaChangeJobV2.java|6|60|10.00%|[523, 524, 525, 533, 534, 535, 536, 537, 538, 540, 541, 543, 544, 545, 546, 548, 549, 551, 552, 556, 557, 558, 559, 561, 566, 567, 568, 569, 571, 572, 573, 574, 576, 578, 580, 582, 584, 585, 586, 587, 588, 589, 590, 593, 596, 597, 598, 599, 600, 602, 604, 605, 606, 607]|
|:large_blue_circle:|com/starrocks/sql/analyzer/AlterTableStatementAnalyzer.java|16|92|17.39%|[104, 305, 306, 309, 311, 312, 314, 315, 317, 320, 321, 322, 323, 324, 326, 329, 330, 331, 332, 333, 334, 335, 337, 338, 340, 343, 344, 345, 346, 348, 349, 354, 355, 356, 357, 358, 359, 360, 375, 428, 429, 430, 431, 433, 435, 454, 455, 458, 460, 461, 463, 464, 466, 469, 470, 471, 472, 473, 475, 478, 479, 480, 481, 482, 483, 484, 486, 487, 489, 492, 493, 494, 495, 497, 498, 512]|
|:large_blue_circle:|com/starrocks/sql/analyzer/UpdateAnalyzer.java|6|33|18.18%|[93, 107, 108, 109, 126, 127, 129, 130, 131, 132, 134, 136, 141, 142, 143, 144, 147, 148, 149, 150, 152, 154, 155, 157, 161, 163, 164]|
|:large_blue_circle:|com/starrocks/alter/SchemaChangeHandler.java|13|48|27.08%|[323, 324, 325, 327, 328, 329, 330, 333, 334, 409, 413, 417, 418, 419, 420, 421, 423, 425, 426, 427, 429, 430, 431, 432, 434, 435, 437, 438, 439, 441, 442, 443, 1216, 1289, 1302]|
|:large_blue_circle:|com/starrocks/sql/InsertPlanner.java|13|40|32.50%|[252, 300, 353, 354, 355, 356, 357, 359, 360, 361, 362, 364, 365, 367, 368, 369, 370, 372, 373, 375, 376, 379, 380, 381, 382, 404, 405]|
|:large_blue_circle:|com/starrocks/load/Load.java|20|59|33.90%|[184, 185, 233, 235, 236, 237, 238, 241, 242, 243, 244, 245, 247, 248, 249, 317, 318, 422, 423, 432, 433, 434, 436, 497, 498, 499, 500, 505, 653, 660, 661, 662, 663, 664, 666, 667, 669, 671, 673]|
|:large_blue_circle:|com/starrocks/sql/parser/AstBuilder.java|15|33|45.45%|[894, 3379, 3380, 3381, 3384, 3385, 3386, 3389, 3390, 3391, 3407, 3408, 3448, 3449, 3450, 3453, 3454, 3455]|
|:large_blue_circle:|com/starrocks/catalog/Column.java|10|21|47.62%|[354, 469, 470, 471, 473, 474, 504, 598, 663, 666, 667]|
|:large_blue_circle:|com/starrocks/sql/ast/ColumnDef.java|5|7|71.43%|[212, 483]|
|:large_blue_circle:|com/starrocks/sql/analyzer/CreateTableAnalyzer.java|44|47|93.62%|[407, 413, 419]|
|:large_blue_circle:|com/starrocks/alter/OlapTableAlterJobV2Builder.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/catalog/OlapTable.java|12|12|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/analyzer/InsertAnalyzer.java|18|18|100.00%|[]|
|:large_blue_circle:|com/starrocks/task/AlterReplicaTask.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/alter/AlterJobV2Builder.java|2|2|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 183 / 478 (38.28%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/ast/AddColumnClause.java|0|2|00.00%|[35, 36]|
|:large_blue_circle:|com/starrocks/alter/SchemaChangeJobV2.java|6|60|10.00%|[523, 524, 525, 533, 534, 535, 536, 537, 538, 540, 541, 543, 544, 545, 546, 548, 549, 551, 552, 556, 557, 558, 559, 561, 566, 567, 568, 569, 571, 572, 573, 574, 576, 578, 580, 582, 584, 585, 586, 587, 588, 589, 590, 593, 596, 597, 598, 599, 600, 602, 604, 605, 606, 607]|
|:large_blue_circle:|com/starrocks/sql/analyzer/AlterTableStatementAnalyzer.java|16|92|17.39%|[104, 305, 306, 309, 311, 312, 314, 315, 317, 320, 321, 322, 323, 324, 326, 329, 330, 331, 332, 333, 334, 335, 337, 338, 340, 343, 344, 345, 346, 348, 349, 354, 355, 356, 357, 358, 359, 360, 375, 428, 429, 430, 431, 433, 435, 454, 455, 458, 460, 461, 463, 464, 466, 469, 470, 471, 472, 473, 475, 478, 479, 480, 481, 482, 483, 484, 486, 487, 489, 492, 493, 494, 495, 497, 498, 512]|
|:large_blue_circle:|com/starrocks/sql/analyzer/UpdateAnalyzer.java|6|34|17.65%|[125, 140, 141, 142, 143, 166, 167, 169, 170, 171, 172, 174, 176, 181, 182, 183, 184, 187, 188, 189, 190, 192, 194, 195, 197, 201, 203, 204]|
|:large_blue_circle:|com/starrocks/alter/SchemaChangeHandler.java|13|48|27.08%|[323, 324, 325, 327, 328, 329, 330, 333, 334, 409, 413, 417, 418, 419, 420, 421, 423, 425, 426, 427, 429, 430, 431, 432, 434, 435, 437, 438, 439, 441, 442, 443, 1216, 1289, 1302]|
|:large_blue_circle:|com/starrocks/sql/InsertPlanner.java|13|40|32.50%|[251, 301, 354, 355, 356, 357, 358, 360, 361, 362, 363, 365, 366, 368, 369, 370, 371, 373, 374, 376, 377, 380, 381, 382, 383, 405, 406]|
|:large_blue_circle:|com/starrocks/load/Load.java|20|59|33.90%|[184, 185, 233, 235, 236, 237, 238, 241, 242, 243, 244, 245, 247, 248, 249, 317, 318, 422, 423, 432, 433, 434, 436, 497, 498, 499, 500, 505, 653, 660, 661, 662, 663, 664, 666, 667, 669, 671, 673]|
|:large_blue_circle:|com/starrocks/sql/parser/AstBuilder.java|15|33|45.45%|[894, 3379, 3380, 3381, 3384, 3385, 3386, 3389, 3390, 3391, 3407, 3408, 3448, 3449, 3450, 3453, 3454, 3455]|
|:large_blue_circle:|com/starrocks/catalog/Column.java|10|21|47.62%|[354, 469, 470, 471, 473, 474, 504, 598, 663, 666, 667]|
|:large_blue_circle:|com/starrocks/sql/ast/ColumnDef.java|5|7|71.43%|[212, 483]|
|:large_blue_circle:|com/starrocks/sql/analyzer/CreateTableAnalyzer.java|44|47|93.62%|[407, 413, 419]|
|:large_blue_circle:|com/starrocks/alter/OlapTableAlterJobV2Builder.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/catalog/OlapTable.java|12|12|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/analyzer/InsertAnalyzer.java|18|18|100.00%|[]|
|:large_blue_circle:|com/starrocks/task/AlterReplicaTask.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/alter/AlterJobV2Builder.java|2|2|100.00%|[]|

File: fe/fe-core/src/main/java/com/starrocks/catalog/Column.java [hunk 20]
```diff
@@ -107,6 +107,10 @@ public class Column implements Writable {
     // Currently, analyzed define expr is only used when creating materialized views, so the define expr in RollupJob must be analyzed.
     // In other cases, such as define expr in `MaterializedIndexMeta`, it may not be analyzed after being relayed.
     private Expr defineExpr; // use to define column in materialize view
+    @SerializedName(value = "isMaterializedColumn")
+    private boolean isMaterializedColumn;
```
[the remark this answers is not in the document]
Comment (chaoyli) [reply fe/fe-core/src/main/java/com/starrocks/catalog/Column.java]: From the ANSI, is `generated column`?

**Review by HangyuanLiu — DISMISSED** [review]

**Review by HangyuanLiu — APPROVED** [review]

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 183 / 480 (38.12%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/ast/AddColumnClause.java|0|2|00.00%|[35, 36]|
|:large_blue_circle:|com/starrocks/alter/SchemaChangeHandler.java|4|36|11.11%|[323, 324, 325, 327, 328, 329, 330, 333, 334, 409, 413, 417, 418, 419, 420, 421, 423, 425, 426, 427, 429, 430, 431, 432, 434, 435, 437, 438, 439, 441, 442, 443]|
|:large_blue_circle:|com/starrocks/sql/analyzer/AlterTableStatementAnalyzer.java|16|92|17.39%|[104, 305, 306, 309, 311, 312, 314, 315, 317, 320, 321, 322, 323, 324, 326, 329, 330, 331, 332, 333, 334, 335, 337, 338, 340, 343, 344, 345, 346, 348, 349, 354, 355, 356, 357, 358, 359, 360, 375, 428, 429, 430, 431, 433, 435, 454, 455, 458, 460, 461, 463, 464, 466, 469, 470, 471, 472, 473, 475, 478, 479, 480, 481, 482, 483, 484, 486, 487, 489, 492, 493, 494, 495, 497, 498, 512]|
|:large_blue_circle:|com/starrocks/sql/analyzer/UpdateAnalyzer.java|6|34|17.65%|[125, 140, 141, 142, 143, 166, 167, 169, 170, 171, 172, 174, 176, 181, 182, 183, 184, 187, 188, 189, 190, 192, 194, 195, 197, 201, 203, 204]|
|:large_blue_circle:|com/starrocks/alter/SchemaChangeJobV2.java|16|72|22.22%|[526, 532, 538, 539, 540, 548, 549, 550, 551, 552, 553, 555, 556, 558, 559, 560, 561, 563, 564, 566, 567, 571, 572, 573, 574, 576, 581, 582, 583, 584, 586, 587, 588, 589, 591, 593, 595, 597, 599, 600, 601, 602, 603, 604, 605, 608, 609, 611, 612, 613, 614, 615, 617, 619, 620, 621]|
|:large_blue_circle:|com/starrocks/sql/InsertPlanner.java|13|42|30.95%|[267, 321, 374, 375, 376, 377, 378, 379, 381, 382, 384, 385, 386, 388, 389, 391, 392, 393, 394, 396, 397, 399, 400, 402, 403, 404, 405, 427, 428]|
|:large_blue_circle:|com/starrocks/load/Load.java|20|60|33.33%|[184, 185, 233, 235, 236, 237, 238, 241, 242, 243, 244, 245, 246, 248, 249, 250, 318, 319, 423, 424, 433, 434, 435, 437, 498, 499, 500, 501, 506, 654, 661, 662, 663, 664, 665, 667, 668, 670, 672, 674]|
|:large_blue_circle:|com/starrocks/sql/parser/AstBuilder.java|15|33|45.45%|[894, 3379, 3380, 3381, 3384, 3385, 3386, 3389, 3390, 3391, 3407, 3408, 3448, 3449, 3450, 3453, 3454, 3455]|
|:large_blue_circle:|com/starrocks/catalog/Column.java|12|23|52.17%|[354, 472, 473, 474, 476, 477, 507, 600, 664, 667, 668]|
|:large_blue_circle:|com/starrocks/sql/ast/ColumnDef.java|5|7|71.43%|[212, 483]|
|:large_blue_circle:|com/starrocks/sql/analyzer/CreateTableAnalyzer.java|44|47|93.62%|[407, 413, 419]|
|:large_blue_circle:|com/starrocks/catalog/OlapTable.java|12|12|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/analyzer/InsertAnalyzer.java|18|18|100.00%|[]|
|:large_blue_circle:|com/starrocks/task/AlterReplicaTask.java|2|2|100.00%|[]|

**Review by chaoyli — APPROVED** [review]

**Pull request merged**